### ISSUE TITLE
[BUG] - type object 'SeqCGenerator' has no attribute 'compress_generators' when number of hashes is too large

### ISSUE THREAD

**Issue by terrorfisch** [zhinst/laboneq#32]
**Describe the bug**
Creating a charge scan fails with an attribute error when the scan is larger than 400 points

**To Reproduce**
```python
x_min, x_max = 0., 1.
y_min, y_max = 0., 1.

t_point = 192 / 2.4e9

x_sweep = LinearSweepParameter(uid="csd_x", start=x_min, stop=x_max, count=20)
y_sweep = LinearSweepParameter(uid="csd_y", start=y_min, stop=y_max, count=20)

constant_pulse = pulse_library.const(length=t_point, amplitude=1.0)

exp_csd = Experiment(
    uid="Amplitude Rabi",
    signals=[
        ExperimentSignal("gate_x"),
        ExperimentSignal("gate_y"),
    ],
)

# shortcut to the logical signal group q0
lsg_0 = device_setup.logical_signal_groups["q0"].logical_signals
lsg_1 = device_setup.logical_signal_groups["q1"].logical_signals

# map signals appropriately
map_signals = {
    "gate_x"  :  lsg_0["plunger"],
    "gate_y"  :  lsg_1["plunger"],
   # "drive-x"  :  lsg["drive_line"],  
}

exp_csd.set_signal_map(map_signals)

with exp_csd.acquire_loop_rt(uid="rabi_shots", count=1, acquisition_type=AcquisitionType.INTEGRATION):
    with exp_csd.sweep(uid="csd_x_sweep", parameter=x_sweep):
        with exp_csd.sweep(uid="csd_y_sweep", parameter=y_sweep):
            with exp_csd.section(uid="csd"):
                exp_csd.play(signal="gate_x", pulse=constant_pulse, amplitude=x_sweep)
                exp_csd.play(signal="gate_y", pulse=constant_pulse, amplitude=y_sweep)

result = session.run(exp_csd)
```

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[5], line 2
      1 # run the experiment
----> 2 result = session.run(exp_csd)

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\dsl\session.py:78, in Session._SessionDeco.entrypoint.<locals>.wrapper(self, *args, **kwargs)
     75 @functools.wraps(func)
     76 def wrapper(self, *args, **kwargs):
     77     self._call_stack.append(func.__name__)
---> 78     res = func(self, *args, **kwargs)
     79     self._call_stack.pop()
     80     return res

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\_observability\tracing\_tracer.py:73, in trace.<locals>.outer_wrapper.<locals>.wrapper(*args, **kwargs)
     71             tracing.enable_tracing()
     72         return res
---> 73 return func(*args, **kwargs)

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\dsl\session.py:298, in Session.run(self, experiment)
    296         self._compiled_experiment = experiment
    297     else:
--> 298         self.compile(experiment)
    300 if not self._assert_connected(fail=False):
    301     return

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\dsl\session.py:78, in Session._SessionDeco.entrypoint.<locals>.wrapper(self, *args, **kwargs)
     75 @functools.wraps(func)
     76 def wrapper(self, *args, **kwargs):
     77     self._call_stack.append(func.__name__)
---> 78     res = func(self, *args, **kwargs)
     79     self._call_stack.pop()
     80     return res

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\_observability\tracing\_tracer.py:73, in trace.<locals>.outer_wrapper.<locals>.wrapper(*args, **kwargs)
     71             tracing.enable_tracing()
     72         return res
---> 73 return func(*args, **kwargs)

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\dsl\session.py:254, in Session.compile(self, experiment, compiler_settings)
    251     return
    253 self._experiment_definition = experiment
--> 254 self._compiled_experiment = LabOneQFacade.compile(
    255     self, self.logger, compiler_settings=compiler_settings
    256 )
    257 self._last_results = None
    258 return self._compiled_experiment

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\dsl\laboneq_facade.py:55, in LabOneQFacade.compile(session, logger, compiler_settings)
     53 logger.debug("Calling LabOne Q Compiler...")
     54 compiler = Compiler.from_user_settings(compiler_settings)
---> 55 compiled_experiment = compiler.run(
     56     {"setup": session.device_setup, "experiment": session.experiment}
     57 )
     58 compiled_experiment.device_setup = session.device_setup
     59 compiled_experiment.experiment = session.experiment

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\_observability\tracing\_tracer.py:73, in trace.<locals>.outer_wrapper.<locals>.wrapper(*args, **kwargs)
     71             tracing.enable_tracing()
     72         return res
---> 73 return func(*args, **kwargs)

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\compiler\workflow\compiler.py:1094, in Compiler.run(self, data)
   1091 self._process_experiment(data)
   1092 self._calc_integration_unit_allocation()
-> 1094 self._generate_code()
   1095 self._generate_recipe()
   1097 retval = self.compiler_output()

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\_observability\tracing\_tracer.py:73, in trace.<locals>.outer_wrapper.<locals>.wrapper(*args, **kwargs)
     71             tracing.enable_tracing()
     72         return res
---> 73 return func(*args, **kwargs)

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\compiler\workflow\compiler.py:372, in Compiler._generate_code(self)
    370 events = self._scheduler.event_timing(expand_loops=False)
    371 code_generator.gen_acquire_map(events, self._experiment_dao)
--> 372 code_generator.gen_seq_c(
    373     events,
    374     {k: self._experiment_dao.pulse(k) for k in self._experiment_dao.pulses()},
    375 )
    376 self._command_table_match_offsets = code_generator.command_table_match_offsets()
    377 self._feedback_connections = code_generator.feedback_connections()

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\compiler\code_generator\code_generator.py:613, in CodeGenerator.gen_seq_c(self, events, pulse_defs)
    605 self._feedback_register_allocator = FeedbackRegisterAllocator(
    606     self._signals, filtered_events
    607 )
    609 for _, awg in sorted(
    610     self._awgs.items(),
    611     key=lambda item: item[0].device_id + str(item[0].awg_number),
    612 ):
--> 613     self._gen_seq_c_per_awg(awg, filtered_events, pulse_defs)
    615 return self._signal_delays

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\compiler\code_generator\code_generator.py:1049, in CodeGenerator._gen_seq_c_per_awg(self, awg, events, pulse_defs)
   1044         seq_c_generators.append(generator)
   1045 _logger.debug(
   1046     "***  collected generators, seq_c_generators: %s", seq_c_generators
   1047 )
-> 1049 main_generator = compress_generators_rle(
   1050     seq_c_generators, declarations_generator
   1051 )
   1053 if main_generator.needs_play_zero_counter():
   1054     declarations_generator.add_variable_declaration(
   1055         main_generator.play_zero_counter_variable_name()
   1056     )

File ~\Documents\laboneq_tests\laboneq_venv\lib\site-packages\laboneq\compiler\code_generator\compressor.py:83, in compress_generators_rle(seq_c_generators, declarations_generator)
     78 if len(hashes_dict) >= 400:
     79     _logger.info(
     80         "The number of hashes %d is too large for compression, using fallback",
     81         len(hashes_dict),
     82     )
---> 83     return SeqCGenerator.compress_generators(
     84         seq_c_generators, declarations_generator
     85     )
     86 as_index_sequence = list(map(lambda x: hashes_dict.index(x), hashes))
     88 _logger.debug(
     89     "The form of the generators is %s", " ".join(map(str, as_index_sequence))
     90 )

AttributeError: type object 'SeqCGenerator' has no attribute 'compress_generators'
```

**Expected behavior**
A working charge scan

**Versions used:**
 - laboneq: `2.3.0`
 - zhinst-core: `23.2.41338`

**Context**
- [ ] Do you know a workaround for the issue? If yes, please provide a short description. 
- [x] Does the issue block your experiments?

**Comment by polw-zi** [comment]
Thanks for the detailed report, @terrorfisch!

I could reproduce your issue. 

We have very recently performed a significant overhaul of that very part of the compiler. I have just checked, your issue is indeed gone in the current develop version of LabOne Q. The fix will be part of version 2.4.

**Comment by QuantumClemens** [comment]
Hi @terrorfisch, 
This was fixed in release 2.4, will close this ticket. Let us know if you still have issues.
